File: fsharp/src/Hangman.fs
module Hangman
    open System
    open WordList
    
    let printStatus currentGuess guessedChars triesLeft: unit = 
        printfn "Word: %s" currentGuess
        printfn "Tries Left: %i" triesLeft
        printfn "Guessed Letters:  %s\n" guessedChars 

        printfn "The Gallow\n\n"

        match triesLeft with
        | 10 -> printfn "\n\n\n\n\n\n"
        | 9 -> printfn "\n\n\n\n\n   \n_______________\n"
        | 8 -> printfn "\n\n\n  |\n  |\n  |\n________________\n"
        | 7 -> printfn "  |\n  |\n  |\n  |\n  |\n  |\n________________\n"
        | 6 -> printfn "  |--         \n  |\n  |\n  |\n  |\n  |\n  |\n________________\n"
        | 5 -> printfn "  |------         \n  |\n  |\n  |\n  |\n  |\n  |\n________________\n"
        | 4 -> printfn "  |------         \n  |     |\n  |\n  |\n  |\n  |\n  |\n________________\n"
        | 3 -> printfn "  |------         \n  |     |\n  |     0\n  |\n  |\n  |\n  |\n________________\n"
        | 2 -> printfn "  |------         \n  |     |\n  |     0\n  |    /|\\\n  |\n  |\n  |\n________________\n"
        | 1 -> printfn "  |------         \n  |     |\n  |     0\n  |    /|\\\n  |     |\n  |\n  |\n________________\n"
        | 0 -> printfn "  |------         \n  |     |\n  |     0\n  |    /|\\\n  |     |\n  |    / \\\n  |\n________________\n"
        | _ -> eprintfn "ERROR: internal error, invalid number of tries left."

    //A string conversion to list, needed as strings cannot be zipped in this version of fsharp
    let stringToList theString: char list = [for c in theString -> c]

    //Created my own version of String.Contains as fsharp's version did not seem to accept chars as input
    let stringContains theChar theString: bool = String.exists (fun c -> c = theChar) theString

    let rec buildUpdatedCurrentGuess guess (wordState: (char*char) list): string =
        match wordState with
        | [] -> ""
        | head :: tail -> 
            if (fst head) = guess
            then Char.ToString (fst head) + buildUpdatedCurrentGuess guess tail
            else Char.ToString (snd head) + buildUpdatedCurrentGuess guess tail

    let buildUpdatedGuessedChars guess guessedChars: string =
        if String.exists (fun c -> c = guess) guessedChars
        then guessedChars
        else guessedChars + Char.ToString guess + " "

    let rec playGame randomWord currentGuess guessedChars triesLeft: bool =
        match triesLeft with
        | 0 -> 
            printStatus currentGuess guessedChars triesLeft
            false
        | _ ->
            if currentGuess = randomWord
            then true
            else
                printStatus currentGuess guessedChars triesLeft

                //Gets a guess from user, and prints a newline for formatting
                //A mutable variable is used, as a function containing ReadLine() runs endlessly on mono or becuase I made a mistake
                let mutable guess = '0'
                while not (Char.IsLetter guess) do
                    printf "Enter your next guess: "
                    let guessString = Console.ReadLine()
                    if guessString.Length = 0
                    then guess <- '0'
                    else guess <- Char.ToLower guessString.[0]
                printfn ""
                               
                //Check if the guess is in the random word, or if the user have already tried it
                if stringContains guess randomWord
                then 
                    let newCurrentGuess = buildUpdatedCurrentGuess guess (List.zip (stringToList randomWord) (stringToList currentGuess))
                    let newGuessedChars = buildUpdatedGuessedChars guess guessedChars
                    playGame randomWord newCurrentGuess newGuessedChars triesLeft
                elif stringContains guess guessedChars 
                then
                    playGame randomWord currentGuess guessedChars triesLeft
                else
                    playGame randomWord currentGuess (buildUpdatedGuessedChars guess guessedChars) (triesLeft-1)

    let startGame wordListPath: unit = 
        let randomWord = getRandomWord wordListPath
        let currentGuess = String.replicate (String.length randomWord) "_"

        if playGame randomWord currentGuess "" 10
        then printfn "Congratulations, you guessed the word \"%s\" correctly." randomWord
        else printfn "You did not guess \"%s\" correctly, and have been hanged." randomWord

File: scala/src/Hangman.scala
class Hangman(wordListPath: String) {
 
  /* Package Methods */
  def startGame(): Unit = {
    //Grabs a new word from the word list and convert it to underscores
    val randomWord = this.wordList.getRandomWord
    val currentGuess = "_" * randomWord.size
   
    //It is much easiere to compare the result if everything is lowercase inside the game
    if(playGame(randomWord.toLowerCase, currentGuess, "", 10))
        println("Congratulations, you guessed the word \"" + randomWord + "\" correctly.")
    else
        println("You did not guess \"" + randomWord + "\" correctly, and have been hanged.")
  }

  /* Private Methods */
  private def playGame(randomWord: String, currentGuess: String, guessedChars: String, tries: Int): Boolean = tries match {
    case 0 => 
      printStatus(currentGuess, guessedChars, tries)
      false
    case _ => {
      if(randomWord == currentGuess)
        return true

      //Prints the status of the game and ask the user for a guess, the println is just for formating of output
      printStatus(currentGuess, guessedChars, tries)
      val guess: Char = getGuess
      println

      //Checks if the used guessed correctly, have tried the letter before, or guessed on a letter not in the word
      if(randomWord.contains(guess)) {
        val newCurrentGuess = buildUpdatedCurrentGuess(guess, randomWord.toList.zip(currentGuess.toList))
        val newGuessedChars = buildUpdatedGuessedChars(guess, guessedChars)
        playGame(randomWord, newCurrentGuess, newGuessedChars, tries)
      }
      else if(guessedChars.contains(guess))
        playGame(randomWord, currentGuess, guessedChars, tries)
      else
        playGame(randomWord, currentGuess, buildUpdatedGuessedChars(guess, guessedChars), tries-1)
      }
  }

  private def getGuess(): Char = {
    try {
        var guess = '0'
        do {
            print("Enter your next guess: ")
            guess = readChar.toLower
        } while (!guess.isLetter)
        guess         
    }
    catch {
      case ex: java.lang.StringIndexOutOfBoundsException => getGuess
    }
  }

  private def buildUpdatedCurrentGuess(guess: Char, wordState: List[(Char, Char)]): String = wordState match {
    case List() => ""
    case x :: xs => if (x._1 == guess) x._1 +: buildUpdatedCurrentGuess(guess, wordState.tail) 
                    else x._2 +: buildUpdatedCurrentGuess(guess, wordState.tail) 
  }

  private def buildUpdatedGuessedChars(guess: Char, guessedChars: String): String = {
    if (guessedChars.contains(guess)) guessedChars
    else guessedChars + " " + guess
  }

  private def printStatus(currentGuess: String, guessedChars: String, tries: Int)
  {
      println("Word: " + currentGuess)
      println("Tries Left: " +tries)
      println("Guessed Letters: "+ guessedChars + "\n")

      println("The Gallow\n\n")

      tries match {
        case 10 => println("\n\n\n\n\n\n")
        case 9 => println("\n\n\n\n\n   \n_______________\n")
        case 8 => println("\n\n\n  |\n  |\n  |\n________________\n")
        case 7 => println("  |\n  |\n  |\n  |\n  |\n  |\n________________\n")
        case 6 => println("  |--         \n  |\n  |\n  |\n  |\n  |\n  |\n________________\n")
        case 5 => println("  |------         \n  |\n  |\n  |\n  |\n  |\n  |\n________________\n")
        case 4 => println("  |------         \n  |     |\n  |\n  |\n  |\n  |\n  |\n________________\n")
        case 3 => println("  |------         \n  |     |\n  |     0\n  |\n  |\n  |\n  |\n________________\n")
        case 2 => println("  |------         \n  |     |\n  |     0\n  |    /|\\\n  |\n  |\n  |\n________________\n")
        case 1 => println("  |------         \n  |     |\n  |     0\n  |    /|\\\n  |     |\n  |\n  |\n________________\n")
        case 0 => println("  |------         \n  |     |\n  |     0\n  |    /|\\\n  |     |\n  |    / \\\n  |\n________________\n")
      }
  }

  /* Instance variables */
  private val wordList: WordList = new WordList(wordListPath)
}
